sys/linux/dev_snd_hw.txt

```diff
@@ -0,0 +1,76 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/asm/ioctl.h>
+include <uapi/linux/fcntl.h>
+include <uapi/linux/time.h>
+include <uapi/sound/asound.h>
+include <uapi/sound/firewire.h>
+
+resource fd_snd_hw[fd]
+
+syz_open_dev$sndhw(dev ptr[in, string["/dev/snd/hwC#D#"]], id intptr, flags flags[open_flags]) fd_snd_hw
+
+read$sndhw(fd fd_snd_hw, buffer ptr[out, array[int8]], count bytesize[buffer])
+write$sndhw(fd fd_snd_hw, buffer ptr[in, array[int8]], count bytesize[buffer])
+
+# This syscall requires actual firewire hardware.
+write$sndhw_fireworks(fd fd_snd_hw, buffer ptr[in, snd_efw_transaction], count bytesize[buffer])
+
+ioctl$SNDRV_HWDEP_IOCTL_PVERSION(fd fd_snd_hw, cmd const[SNDRV_HWDEP_IOCTL_PVERSION], arg ptr[out, int32])
+ioctl$SNDRV_HWDEP_IOCTL_INFO(fd fd_snd_hw, cmd const[SNDRV_HWDEP_IOCTL_INFO], arg ptr[out, snd_hwdep_info])
+ioctl$SNDRV_HWDEP_IOCTL_DSP_STATUS(fd fd_snd_hw, cmd const[SNDRV_HWDEP_IOCTL_DSP_STATUS], arg ptr[out, snd_hwdep_dsp_status])
+ioctl$SNDRV_HWDEP_IOCTL_DSP_LOAD(fd fd_snd_hw, cmd const[SNDRV_HWDEP_IOCTL_DSP_LOAD], arg ptr[in, snd_hwdep_dsp_image])
+
+# These ioctls require actual firewire hardware.
+ioctl$SNDRV_FIREWIRE_IOCTL_GET_INFO(fd fd_snd_hw, cmd const[SNDRV_FIREWIRE_IOCTL_GET_INFO], arg ptr[out, snd_firewire_get_info])
+ioctl$SNDRV_FIREWIRE_IOCTL_LOCK(fd fd_snd_hw, cmd const[SNDRV_FIREWIRE_IOCTL_LOCK])
+ioctl$SNDRV_FIREWIRE_IOCTL_UNLOCK(fd fd_snd_hw, cmd const[SNDRV_FIREWIRE_IOCTL_UNLOCK])
+ioctl$SNDRV_FIREWIRE_IOCTL_TASCAM_STATE(fd fd_snd_hw, cmd const[SNDRV_FIREWIRE_IOCTL_TASCAM_STATE], arg ptr[out, snd_firewire_tascam_state])
+
+snd_hwdep_info {
+	device		int32
+	card		int32
+	id		array[int8, 64]
+	name		array[int8, 80]
+	iface		int32[SNDRV_HWDEP_IFACE_OPL2:SNDRV_HWDEP_IFACE_LAST]
+	reserved	array[int8, 64]
+}
+
+snd_hwdep_dsp_status {
+	version		int32
+	id		array[int8, 32]
+	num_dsps	int32
+	dsp_loaded	int32
+	chip_ready	int32
+	reserved	array[int8, 16]
+}
+
+snd_hwdep_dsp_image {
+	index		int32[0:31]
+	name		array[int8, 64]
+	image		ptr[in, array[int8]]
+	length		bytesize[image, int32]
```

Comment:
syz-check says:
```
bad-struct-size: snd_hwdep_dsp_image: syz=84 kernel=80 [386]
bad-struct-size: snd_hwdep_dsp_image: syz=88 kernel=80 [arm]
bad-field-size: snd_hwdep_dsp_image.length: syz=4 kernel=8 [amd64,arm64]
bad-field-offset: snd_hwdep_dsp_image.driver_data: syz=80 kernel=76 [arm]
bad-field-size: snd_hwdep_dsp_image.driver_data: syz=8 kernel=4 [386,arm]
```
length is size_t and driver_data unsigned long

Reply:
I understand for driver_data but for size_t, I don't understand what is wrong. Maybe I need to specify this structure as packed?

Reply:
I did some fixes. Are we still getting same errors?

Reply:
size_t is a typedef for `unsigned long`, pointer-sized integer. You used `int32` which is always 4 bytes, whereas `unsigned long` is 8 bytes on 64-bit arches.